experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,697 @@
+            bool &outsValid) {
+    outsValid = true;
+    for (unsigned i = 0; i < size; ++i) {
+      outsValid = outsValid && insValid[i];
+    }
+
+    for (unsigned i = 0; i < size; ++i) {
+      insReady[i].get() = outsReady;
+      for (unsigned j = 0; j < size; ++j) {
+        if (i != j) {
+          insReady[i].get() = insReady[i].get() && insValid[j];
+        }
+      }
+    }
+  }
+
+private:
+  unsigned size;
+};
+
+class DelayBuffer {
+public:
+  DelayBuffer(unsigned size = 32) : size(size) {}
+  void exec(bool isReset, bool validIn, bool readyIn, bool &validOut) {
+    if (isReset) {
+      if (size == 1) {
+        validOut = validIn;
+      } else {
+        validOut = false;
+      }
+    } else if (readyIn) {
+      validOut = validIn;
+    }
+  }
+
+private:
+  unsigned size;
+};
+
+class OEHB {
+public:
+  OEHB(unsigned dataWidth) : dataWidth(dataWidth) {}
+
+  void exec(bool isReset, const Any &ins, bool insValid, bool outsReady,
+            bool &insReady, Any &outs, bool &outsValid) {
+    if (isReset) {
+      outsValid = false;
+      insReady = true;
+      outs = 0;
+    } else {
+      outsValid = insValid;
+      insReady = !insValid || outsReady;
+      if (insValid && outsReady) {
+        outs = ins;
+      }
+    }
+  }
+
+private:
+  unsigned dataWidth;
+};
+
+class TEHBDataless {
+public:
+  TEHBDataless() = default;
+  void exec(bool isReset, bool insValid, bool outsReady, bool &insReady,
+            bool &outsValid) {
+    if (isReset) {
+      insReady = 1;
+    } else {
+      insReady = !insValid || outsReady;
+    }
+    outsValid = insValid;
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Cocrete Execution Models
+//===----------------------------------------------------------------------===//
+class SinkModel : public OpExecutionModel<handshake::SinkOp> {
+public:
+  using OpExecutionModel<handshake::SinkOp>::OpExecutionModel;
+  SinkModel(handshake::SinkOp sinkOp)
+      : OpExecutionModel<handshake::SinkOp>(sinkOp), op(sinkOp) {}
+
+  void exec(bool isReset,
+            mlir::DenseMap<std::pair<Value, Operation *>, RW *> &rws) override {
+    auto *ins = getState<ControlConsumerRW>(rws[{op.getOperand(), op}]);
+    ins->ready = true;
+  }
+
+private:
+  handshake::SinkOp op;
+};
+
+class MulIModel : public OpExecutionModel<handshake::MulIOp> {
+public:
+  using OpExecutionModel<handshake::MulIOp>::OpExecutionModel;
+  MulIModel(handshake::MulIOp muliOp)
+      : OpExecutionModel<handshake::MulIOp>(muliOp), op(muliOp),
+        bitwidth(cast<handshake::ChannelType>(muliOp.getLhs().getType())
+                     .getDataBitWidth()),
+        muliJoin(2), muliMul4Stage(bitwidth), muliDelayBuffer(3),
+        muliOEHB(bitwidth) {}
+
+  void exec(bool isReset,
+            mlir::DenseMap<std::pair<Value, Operation *>, RW *> &rws) override {
+    auto *lhs = getState<ChannelConsumerRW>(rws[{op.getLhs(), op}]);
```

Comment:
Wouldn't it be nicer to do all this state retrieval (these first 5 lines) once in the constructor and store pointers/references to all your state readers/writer in private class members? It seems to make semantic sense because none of it depends on the current execution state.